include/quic/parser.hpp

```diff
@@ -0,0 +1,421 @@
+
+        int64_t rid() { return req_id; }
+        std::string_view view() { return {data}; }
+    };
+
+    struct sent_request
+    {
+        // parsed request data
+        int64_t req_id;
+        std::string data;
+        bparser& return_sender;
+
+        // total length of the request; is at the beginning of the request
+        size_t total_len;
+
+        std::chrono::steady_clock::time_point req_time;
+        std::chrono::steady_clock::time_point timeout;
+
+        bool is_empty() const { return data.empty() && total_len == 0; }
+
+        explicit sent_request(bparser& bp, std::string d, int64_t rid) : req_id{rid}, return_sender{bp}
+        {
+            total_len = d.length();
+            data.reserve(data.length() + total_len);
+            data += std::to_string(total_len);
+            data += ':';
+            data.append(d);
+
+            req_time = get_time();
+            timeout = req_time + TIMEOUT;
+        }
+
+        bool is_expired(std::chrono::steady_clock::time_point tp) { return timeout < tp; }
+
+        message to_message(bool timed_out = false) { return {return_sender, data, timed_out}; }
+
+        std::string_view view() { return {data}; }
+        std::string&& payload() { return std::move(data); }
+    };
+
+    class bparser : public Stream
+    {
+      private:
+        // outgoing requests awaiting response
+        std::deque<std::shared_ptr<sent_request>> sent_reqs;
+
+        std::string buf;
+        std::string size_buf;
+
+        size_t current_len{0};
+
+        std::atomic<int64_t> next_rid{0};
+
+        friend class sent_request;
+        std::function<void(message)> recv_callback;
+
+      public:
+        template <typename... Opt>
+        explicit bparser(Connection& _c, Endpoint& _e, Opt&&... opts) : Stream{_c, _e}
+        {
+            ((void)handle_bp_opt(std::forward<Opt>(opts)), ...);
+        }
+
+        ~bparser() { sent_reqs.clear(); }
+
+        std::weak_ptr<bparser> weak_from_this() { return std::dynamic_pointer_cast<bparser>(shared_from_this()); }
+
+        void request(std::string endpoint, std::string body)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_request(std::move(endpoint), std::move(body));
+            send(req->view());
+
+            sent_reqs.push_back(std::move(req));
+        }
+
+        void command(std::string endpoint, std::string body)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_command(std::move(endpoint), std::move(body));
+            send(req->payload());
+        }
+
+        void respond(int64_t rid, std::string body, bool error = false)
+        {
+            log::debug(bp_cat, "{} called", __PRETTY_FUNCTION__);
+
+            auto req = make_response(rid, std::move(body), error);
+            send(req->payload());
+        }
+
+        void check_timeouts()
+        {
+            const auto& now = get_time();
```

Comment:
```suggestion
            const auto now = get_time();
```
I think this was a reference to a temporary